verilog/analysis/checkers/always_ff_non_blocking_rule.cc

```diff
@@ -0,0 +1,67 @@
+// Copyright 2017-2019 The Verible Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "verilog/analysis/checkers/always_ff_non_blocking_rule.h"
+
+#include <string>
+#include <vector>
+
+#include "absl/strings/str_cat.h"
+#include "absl/strings/string_view.h"
+#include "common/analysis/citation.h"
+#include "common/analysis/lint_rule_status.h"
+#include "common/analysis/matcher/bound_symbol_manager.h"
+#include "common/text/symbol.h"
+#include "common/text/syntax_tree_context.h"
+#include "verilog/analysis/descriptions.h"
+#include "verilog/analysis/lint_rule_registry.h"
+
+namespace verilog {
+namespace analysis {
+
+using verible::GetStyleGuideCitation;
+using verible::LintRuleStatus;
+using verible::LintViolation;
+using verible::SyntaxTreeContext;
+
+// Register AlwaysFFNonBlockingRule
+VERILOG_REGISTER_LINT_RULE(AlwaysFFNonBlockingRule);
+
+absl::string_view AlwaysFFNonBlockingRule::Name() { return "always-ff-non-blocking"; }
+const char AlwaysFFNonBlockingRule::kTopic[] = "sequential-logic";
+const char AlwaysFFNonBlockingRule::kMessage[] =
+    "Use non-blocking assignment instead of blocking one.";
+
+std::string AlwaysFFNonBlockingRule::GetDescription(DescriptionType description_type) {
+  return absl::StrCat("Checks that there are no occurrences of ",
+                      "blocking assignment in sequential logic.");
+}
+
+void AlwaysFFNonBlockingRule::HandleSymbol(const verible::Symbol& symbol,
+                                  const SyntaxTreeContext& context) {
+  verible::matcher::BoundSymbolManager manager;
+  if (always_ff_matcher_.Matches(symbol, &manager)) {
+    violations_.push_back(LintViolation(symbol, kMessage, context));
+  }
+  if (always_ff_block_matcher_.Matches(symbol, &manager)) {
+    violations_.push_back(LintViolation(symbol, kMessage, context));
+  }
```

Comment:
Similar suggestion to the always_comb implementation: consider using a SearchSyntaxTree once you've matched always_ff_matcher_.
Again, only report on the individual violating assignments instead of the always_ff keyword.

Reply:
Matched always_ff_matcher and then found and reported assignment operator.